components/sup/src/manager/service/supervisor.rs

```diff
@@ -58,7 +59,13 @@ impl Supervisor {
             Some(pid) => Some(pid),
             None => {
                 if self.pid_file.exists() {
-                    Some(read_pid(&self.pid_file).unwrap())
+                    match read_pid(&self.pid_file) {
+                        Ok(pid) => Some(pid),
+                        Err(e) => {
+                            error!("Error reading PID file: {}", e);
+                            None
+                        }
+                    }
```

Comment:
This is the only caller for `read_pid` and it doesn't propagate the error, so if we move the error printing into `read_pid`, and have it return an `Option` we reduce `match` statement to:
```rust
                    read_pid(&self.pid_file)
```

Further, since L59 is redundant, we can reduce the whole `let pid` block starting on L58 to:
```rust
        let pid = self.pid.or_else(|| {
                              if self.pid_file.exists() {
                                  read_pid(&self.pid_file)
                              } else {
                                  None
                              }
                          });
```

And if we move the `self.pid_file.exists()` check inside `read_pid` (we could even avoid the second syscall with a match arm like:
```rust
        // Don't log an error if the file doesn't exist
        Err(ref err) if err.kind() == std::io::ErrorKind::NotFound => return None,
```
), we further simplify to:
```rust
        let pid = self.pid.or_else(|| read_pid(&self.pid_file));
```

But since we do an `if let Some(pid) = pid` directly afterward on L74, why not combine with `and_then`? That gets a little deep with the `process::is_alive` condition and need to return, but if we pull that out and recognize that we always assign `self.pid`, we get something quite nice:
```rust
        self.pid = self.pid
                       .or_else(|| read_pid(&self.pid_file))
                       .and_then(|pid| {
                           if process::is_alive(pid) {
                               Some(pid)
                           } else {
                               debug!("Could not find a live process with pid {:?}", pid);
                               None
                           }
                       });

        if self.pid.is_some() {
            self.change_state(ProcessState::Up);
        } else {
            self.change_state(ProcessState::Down);
            self.cleanup_pidfile();
        }
        self.pid.is_some()
```

I *believe* that's equivalent with the exception that you'll never see `Could not find a live process with pid None` in debug logs. I realize that's a pretty major refactoring, so take or leave what you like.

Reply:
This is very similar to what I was kicking around the other day on another PR, as a clarification of the `check_process` function, completely separate from this `unwrap` issue.

Funnily enough, a big thing that spurred me to start poking around here was noticing that the error message about not finding a live process with the PID `None` was nonsense :laughing: So I'm totally fine with the fact that we don't emit that message.

I think this is equivalent in behavior, otherwise.

Reply:
Nice, that's a great improvement.